### PR TITLE
Feat/bid validation

### PR DESCRIPTION
**Pull request by giorgosera**

### CODE REVIEW

File: lib/openrtb2_3/bid.js [hunk 1]
```diff
@@ -41,7 +42,30 @@ Bid.prototype.replaceMacros = function(string, valuesMap){
...
+Bid.prototype.hasAdm = function() {
+  // Double ! is intentional
+  return !!this.adm && typeof this.adm === 'string';
+};
+
+Bid.prototype.isBlockedAdomain = function(badv) {
+  if (!Array.isArray(badv)) {
+    throw new Error('badv should be an array');
+  }
+  return badv.some(a => ~this.adomain.indexOf(a));
```
Comment (emiliosnic) [inline lib/openrtb2_3/bid.js:58]: _Question_ Is `this.adomain` guaranteed to be an array ?

File: lib/openrtb2_3/bid.js [hunk 2]
```diff
@@ -41,7 +42,30 @@ Bid.prototype.replaceMacros = function(string, valuesMap){
...
+};
+
+Bid.prototype.isBlockedCat = function(bcat) {
+  if (!Array.isArray(bcat)) {
+    throw new Error('bcat should be an array');
+  }
+  return bcat.some(c => ~this.cat.indexOf(c));
+};
+
+const BidBuilder = function(){};
```
Comment (sainis-p) [inline lib/openrtb2_3/bid.js:68]: I like that we moved the validation checks as part of the Bid prototype! 👍

**Comment by sainis-p** [comment]
pope-ship-it

File: lib/openrtb2_3/bid.js (hunk 1 above)
Comment (sainis-p) [inline lib/openrtb2_3/bid.js:58]: Yes by the rtb protocol.
http://www.iab.com/wp-content/uploads/2015/05/OpenRTB_API_Specification_Version_2_3_1.pdf
Page 30

Comment (giorgosera) [inline lib/openrtb2_3/bid.js:58]: No and it's a good point.

File: lib/openrtb2_3/bid.js (hunk 2 above)
Comment (giorgosera) [inline lib/openrtb2_3/bid.js:68]: I like it too:) It feels right. I wanted to move the native asset validation here too but I don't want to deviate from priorities for now.

File: lib/openrtb2_3/bid.js (hunk 1 above)
Comment (sainis-p) [inline lib/openrtb2_3/bid.js:58]: Do we not validate the bidResponse based on rtb specifications before that state? I thought we did? My bad.

Comment (giorgosera) [inline lib/openrtb2_3/bid.js:58]: @gadget69 you are right so yes it's guaranteed. My bad after all :) So no problem with this.

**Pull request merged**